### ISSUE TITLE
api.nextcloud.com (sometimes?) presents invalid certificate.

### ISSUE THREAD

**Issue by noci2012** [nextcloud/server#19783]
The certificate presented is for www.nextcloud.com ==> actions refering to api.nextcloud.com fail.
(f.e. installing new modules seem to fail and cause the following log message).

this is the result of updating V17.0.2.1 -> V18.0.1.

Imho this is a serious security issue of high priority. 

[appstoreFetcher] Warning: GuzzleHttp\Exception\RequestException: cURL error 60: SSL: no alternative certificate subject name matches target host name 'api.nextcloud.com' (see http://curl.haxx.se/libcurl/c/libcurl-errors.html) at <<closure>>

 0. /data/WWW/www.example.net/htdocs/3rdparty/guzzlehttp/guzzle/src/Handler/CurlFactory.php line 149
    GuzzleHttp\Handler\CurlFactory::createRejection(GuzzleHttp\Handl ... l}, {errno: 60,error ... 2})
 1. /data/WWW/www.example.net/htdocs/3rdparty/guzzlehttp/guzzle/src/Handler/CurlFactory.php line 102
    GuzzleHttp\Handler\CurlFactory::finishError(GuzzleHttp\Handler\CurlHandler {}, GuzzleHttp\Handl ... l}, GuzzleHttp\Handler\CurlFactory {})
 2. /data/WWW/www.example.net/htdocs/3rdparty/guzzlehttp/guzzle/src/Handler/CurlHandler.php line 43
    GuzzleHttp\Handler\CurlFactory::finish(GuzzleHttp\Handler\CurlHandler {}, GuzzleHttp\Handl ... l}, GuzzleHttp\Handler\CurlFactory {})
 3. /data/WWW/www.example.net/htdocs/3rdparty/guzzlehttp/guzzle/src/Handler/Proxy.php line 28
    GuzzleHttp\Handler\CurlHandler->__invoke("*** sensitive parameter replaced ***", "*** sensitive parameter replaced ***")
 4. /data/WWW/www.example.net/htdocs/3rdparty/guzzlehttp/guzzle/src/Handler/Proxy.php line 51
    GuzzleHttp\Handler\Proxy::GuzzleHttp\Handler\{closure}("*** sensitive parameters replaced ***")
 5. /data/WWW/www.example.net/htdocs/3rdparty/guzzlehttp/guzzle/src/PrepareBodyMiddleware.php line 37
    GuzzleHttp\Handler\Proxy::GuzzleHttp\Handler\{closure}("*** sensitive parameters replaced ***")
 6. /data/WWW/www.example.net/htdocs/3rdparty/guzzlehttp/guzzle/src/Middleware.php line 30
    GuzzleHttp\PrepareBodyMiddleware->__invoke("*** sensitive parameter replaced ***", "*** sensitive parameter replaced ***")
 7. /data/WWW/www.example.net/htdocs/3rdparty/guzzlehttp/guzzle/src/RedirectMiddleware.php line 70
    GuzzleHttp\Middleware::GuzzleHttp\{closure}("*** sensitive parameters replaced ***")
 8. /data/WWW/www.example.net/htdocs/3rdparty/guzzlehttp/guzzle/src/Middleware.php line 60
    GuzzleHttp\RedirectMiddleware->__invoke("*** sensitive parameter replaced ***", "*** sensitive parameter replaced ***")
 9. /data/WWW/www.example.net/htdocs/3rdparty/guzzlehttp/guzzle/src/HandlerStack.php line 67
    GuzzleHttp\Middleware::GuzzleHttp\{closure}("*** sensitive parameters replaced ***")
10. /data/WWW/www.example.net/htdocs/3rdparty/guzzlehttp/guzzle/src/Client.php line 277
    GuzzleHttp\HandlerStack->__invoke("*** sensitive parameter replaced ***", "*** sensitive parameter replaced ***")
11. /data/WWW/www.example.net/htdocs/3rdparty/guzzlehttp/guzzle/src/Client.php line 125
    GuzzleHttp\Client->transfer("*** sensitive parameter replaced ***", "*** sensitive parameter replaced ***")
12. /data/WWW/www.example.net/htdocs/3rdparty/guzzlehttp/guzzle/src/Client.php line 131
    GuzzleHttp\Client->requestAsync("get", GuzzleHttp\Psr7\Uri {}, {verify: "/data/ ... }})
13. /data/WWW/www.example.net/htdocs/lib/private/Http/Client/Client.php line 149
    GuzzleHttp\Client->request("get", "https://api.nextcloud.com/v1/apps.json", {proxy: null,ver ... e})
14. /data/WWW/www.example.net/htdocs/lib/private/App/AppStore/Fetcher/Fetcher.php line 109
    OC\Http\Client\Client->get("https://api.nextcloud.com/v1/apps.json", {timeout: 10})
15. /data/WWW/www.example.net/htdocs/lib/private/App/AppStore/Fetcher/AppFetcher.php line 86
    OC\App\AppStore\Fetcher\Fetcher->fetch("", "")
16. /data/WWW/www.example.net/htdocs/lib/private/App/AppStore/Fetcher/Fetcher.php line 173
    OC\App\AppStore\Fetcher\AppFetcher->fetch("", "")
17. /data/WWW/www.example.net/htdocs/lib/private/Installer.php line 401
    OC\App\AppStore\Fetcher\Fetcher->get()
18. /data/WWW/www.example.net/htdocs/apps/settings/lib/Controller/AppSettingsController.php line 255
    OC\Installer->isUpdateAvailable("survey_client")
19. <<closure>>
    OCA\Settings\Controller\AppSettingsController->OCA\Settings\Controller\{closure}("*** sensitive parameters replaced ***")
20. /data/WWW/www.example.net/htdocs/apps/settings/lib/Controller/AppSettingsController.php line 248
    array_map(Closure {}, {survey_client:  ... }})
21. /data/WWW/www.example.net/htdocs/lib/private/AppFramework/Http/Dispatcher.php line 170
    OCA\Settings\Controller\AppSettingsController->listApps()
22. /data/WWW/www.example.net/htdocs/lib/private/AppFramework/Http/Dispatcher.php line 99
    OC\AppFramework\Http\Dispatcher->executeController(OCA\Settings\Con ... {}, "listApps")
23. /data/WWW/www.example.net/htdocs/lib/private/AppFramework/App.php line 125
    OC\AppFramework\Http\Dispatcher->dispatch(OCA\Settings\Con ... {}, "listApps")
24. /data/WWW/www.example.net/htdocs/lib/private/AppFramework/Routing/RouteActionHandler.php line 47
    OC\AppFramework\App::main("OCA\\Settings\\ ... r", "listApps", OC\AppFramework\ ... {}, {_route: "settings.AppSettings.listApps"})
25. <<closure>>
    OC\AppFramework\Routing\RouteActionHandler->__invoke({_route: "settings.AppSettings.listApps"})
26. /data/WWW/www.example.net/htdocs/lib/private/Route/Router.php line 299
    call_user_func(OC\AppFramework\ ... {}, {_route: "settings.AppSettings.listApps"})
27. /data/WWW/www.example.net/htdocs/lib/base.php line 1008
    OC\Route\Router->match("/settings/apps/list")
28. /data/WWW/www.example.net/htdocs/index.php line 38
    OC::handleRequest()

GET /settings/apps/list

**Comment by kesselb** [comment]
cc @blizzz

**Comment by noci2012** [comment]
IP =  95.217.53.153
An hour ago apparantly access was possible..., now it is not   so maybe some servers running in rotation or behind LB's have a problem.

**Comment by noci2012** [comment]
It changed now it presents: www.nextcloud.at:

*   Trying 2a01:4f9:2b:29dc::153:443...
* TCP_NODELAY set
* Connected to api.nextcloud.com (2a01:4f9:2b:29dc::153) port 443 (#0)
* ALPN, offering h2
* ALPN, offering http/1.1
* successfully set certificate verify locations:
*   CAfile: /etc/ssl/certs/ca-certificates.crt
  CApath: /etc/ssl/certs
* TLSv1.3 (OUT), TLS handshake, Client hello (1):
* TLSv1.3 (IN), TLS handshake, Server hello (2):
* TLSv1.2 (IN), TLS handshake, Certificate (11):
* TLSv1.2 (IN), TLS handshake, Server key exchange (12):
* TLSv1.2 (IN), TLS handshake, Server finished (14):
* TLSv1.2 (OUT), TLS handshake, Client key exchange (16):
* TLSv1.2 (OUT), TLS change cipher, Change cipher spec (1):
* TLSv1.2 (OUT), TLS handshake, Finished (20):
* TLSv1.2 (IN), TLS handshake, Finished (20):
* SSL connection using TLSv1.2 / ECDHE-RSA-AES256-GCM-SHA384
* ALPN, server accepted to use http/1.1
* Server certificate:
*  subject: CN=www.nextcloud.at
*  start date: Feb  8 04:00:33 2020 GMT
*  expire date: May  8 04:00:33 2020 GMT
*  issuer: C=US; O=Let's Encrypt; CN=Let's Encrypt Authority X3
*  SSL certificate verify ok.
> GET / HTTP/1.1
> Host: api.nextcloud.com
> User-Agent: curl/7.68.0
> Accept: */*
>

**Comment by noci2012** [comment]
@kesselb i doubt i am the only one.. with issues.  and activation of a app gives an error.
Due to upgrade some of the "Untrusted" but working apps were disabled, re-enabling is an issue.
(f.e. talk cannot be reinstated). 

Unless ofcourse talk is not of any real value here.

**Comment by kesselb** [comment]
> Is nextcloud 18 actually ready for stable?

What makes you think that Nextcloud 17 or 16 use a different endpoint for the app store? They are also affected by a broken certificate.

cc @nextcloud/infrastructure

**Comment by noci2012** [comment]
Nextcloud  18 broke before (when the announcement was let out) for some reason this worked in Nextcloud 17 (just before the upgrade), upgrade done, reinstating some disabled apps and it doesn't work anymore seems to hint at a problem that started during upgrade (within a period of about one hour).

**Comment by kesselb** [comment]
If you change the app store url to https://api.nextcloud.com/v1/apps.json fine by me. But that's just the wrong endpoint. Please don't use the issue tracker as support channel.

**Comment by tflidd** [comment]
@kesselb why did you close this? I still get the certificate error and sslLabs also shows a certificate name mismatch: https://www.ssllabs.com/ssltest/analyze.html?d=api.nextcloud.com
![grafik](https://user-images.githubusercontent.com/8111200/76140789-2f997580-605e-11ea-9a7a-0f3591524d3c.png)

**Comment by noci2012** [comment]
for me case closed...
For some reason that url was in my config as 
> appstoreurl => https://api.nextcloud.com/v1
removing the line made it possible to update again.
I have no recollection of ever adding that line.. 
AFAIK i once updated the database password by hand.

**Comment by noci2012** [comment]
wrt. the certificate: why not use a wildcard certificate for the servers?
If DNS updates are a hassle, then look into the standalone dns authenticator.
It can handle a static DNS setup with the certbot running as a "DNS" server during the update.

**Comment by kesselb** [comment]
> why did you close this?

```
curl https://api.nextcloud.com/v1/apps.json
curl: (60) SSL: no alternative certificate subject name matches target host name 'api.nextcloud.com'
```

1. Let's do a curl request for the url from stack trace.

```
curl -k https://api.nextcloud.com/v1/apps.json
<!DOCTYPE HTML PUBLIC "-//IETF//DTD HTML 2.0//EN">
<html><head>
<title>302 Found</title>
</head><body>
<h1>Found</h1>
<p>The document has moved <a href="https://nextcloud.com/de/v1/apps.json">here</a>.</p>
<hr>
<address>Apache/2.4.29 (Ubuntu) Server at api.nextcloud.com Port 443</address>
</body></html>
```

2. Let's do a curl request but allow insecure server connections (`-k`). That seems to be a redirect to nextcloud.com.

https://github.com/nextcloud/server/blob/1bf44e1d973593014e63095f3e7d1348af45b52a/lib/private/App/AppStore/Fetcher/Fetcher.php#L224-L226

3. Check the code. See the default endpoint for the app store is a different url. Close it as configuration issue.

**Comment by noci2012** [comment]
For https any connect is required to be correct before any data is transferred. A redirect happens after a transfer has been done. So this setup will never work. to redirect will never work for anything with non matching hostnames in URL & certificates.
I agree it is a configuration issue  on both sides of the connection.

**Comment by noci2012** [comment]
This may need adjusting a little
https://docs.nextcloud.com/server/18/admin_manual/apps_management.html?highlight=appstore#using-your-own-appstore 

By f.e. using apps.example.com ??

**Comment by kesselb** [comment]
https://github.com/nextcloud/documentation/pull/1868

**Comment by noci2012** [comment]
ok thanks.

**Comment by tflidd** [comment]
>     1. Let's do a curl request for the url from stack trace.
> 
> 
> ```
> curl -k https://api.nextcloud.com/v1/apps.json
> <!DOCTYPE HTML PUBLIC "-//IETF//DTD HTML 2.0//EN">
> <html><head>
> <title>302 Found</title>
> </head><body>

Well, 302 is a temporary redirect, so perhaps Nextcloud needs more resources during roll-out of new release and redirects.
Use a permanent and update the value in the config-file, that doesn't disturb users with strange errors. Or check on the admin page if the old URL is still used.

> ```
> 
>     1. Let's do a curl request but allow insecure server connections (`-k`). That seems to be a redirect to nextcloud.com.

Insecure server connections for an app store, I hope you are joking...



> nextcloud/documentation#1868

Thanks for that.

**Comment by kesselb** [comment]
You asked me why I closed a issue. As response I explained to you step by step how I came to the conclusion that this is a configuration issue.

I have not suggested to disable secure connections for the apps store. Yes the certificate for `api.nextcloud.com` is invalid. But that's completely irrelevant because even with a valid certificate it would not work because the apps store url is a different one. 

The issue here was that he added `'appstoreurl' => 'https://api.nextcloud.com/v1',` to `config.php` and then no connection was possible. 

That's like going to Media Markt and asking for bananas and then complaining the Media Markt does not sell bananas.

I'm going to unsubscribe from this issue now. Probably ping someone else if you think there a things left to discuss or a reopen is required.
